columnflow/calibration/cms/egamma.py

```diff
@@ -0,0 +1,581 @@
+        # one correction is applied in a row
+        pt_eval = flat_np_view(events[self.source_field].rawPt, axis=1)
+
+        # the final corrections must be applied to the current pt though
+        pt_application = flat_np_view(events[self.source_field].pt, axis=1)
+
+        broadcasted_run = ak.broadcast_arrays(
+            events[self.source_field].pt, events.run,
+        )
+        run = flat_np_view(broadcasted_run[1], axis=1)
+        gain = flat_np_view(events[self.source_field].seedGain, axis=1)
+        sceta = flat_np_view(events[self.source_field].superclusterEta, axis=1)
+        r9 = flat_np_view(events[self.source_field].r9, axis=1)
+
+        # prepare arguments
+        # we use pt as et since there depends in linear (following the recoomendations)
+        # (energy is part of the LorentzVector behavior)
+        variable_map = {
+            "et": pt_eval,
+            "eta": sceta,
+            "gain": gain,
+            "r9": r9,
+            "run": run,
+            **self.scale_config.corrector_kwargs,
+        }
+        args = tuple(
+            variable_map[inp.name] for inp in self.scale_corrector.inputs
+            if inp.name in variable_map
+        )
+
+        # varied corrections are only applied to MC
+        if self.with_uncertainties and self.dataset_inst.is_mc:
+            scale_uncertainties = self.scale_corrector("total_uncertainty", *args)
+            scales_up = (1 + scale_uncertainties)
+            scales_down = (1 - scale_uncertainties)
+
+            for (direction, scales) in [("up", scales_up), ("down", scales_down)]:
+                # copy pt and mass
+                pt_varied = ak_copy(events[self.source_field].pt)
+                pt_view = flat_np_view(pt_varied, axis=1)
+
+                # apply the scale variation
+                pt_view *= scales
+
+                # save columns
+                postfix = f"scale_{direction}"
+                events = set_ak_column_f32(
+                    events, f"{self.source_field}.pt_{postfix}", pt_varied,
+                )
+
+        # apply the nominal correction
+        # note: changes are applied to the views and directly propagate to the original ak arrays
+        # and do not need to be inserted into the events chunk again
+        # EGamma energy correction is ONLY applied to DATA
+        if self.dataset_inst.is_data:
+            scales_nom = self.scale_corrector("total_correction", *args)
+            pt_application *= scales_nom
+
+        return events
+
+    def init_func(self: Calibrator) -> None:
+        self.uses |= {
+            # nano columns
+            f"{self.source_field}.{{seedGain,pt,superclusterEta,r9}}",
+            "run",
+            optional_column(f"{self.source_field}.rawPt"),
+        }
+        self.produces |= {
+            f"{self.source_field}.pt",
+            optional_column(f"{self.source_field}.rawPt"),
+        }
+        self.scale_config = self.get_scale_config()
```

Comment:
Can you move this assignment to the setup()? I guess it's only needed within the run() function.

Reply:
Good catch, I copied this from the jets calibrators, who actually need the configs in their init. Fixed :+1: